Use a per-hcp router for all hcp ingress traffic

alvaroaleman opened the pull request:
This change adds a new mode to the cpo where it will set up a router and use it for all controlplane traffic, ref: https://issues.redhat.com/browse/HOSTEDCP-485

This is done by:
* Making the CLI decide based on the presence of a value for `--external-dns-domain` if it configures Route or the existing LoadBalancer as publishingStrategy for the KAS
* If the KAS publishing strategy is route, the CPO will:
    * Not set up a LB but a clusterIP service for the KAS
    * Create an external route if the cluster is public
    * Create an internal route
    * Configure the router with a custom template that uses the KAS as default backend, this is needed because in-cluster traffic accesses the KAS through an IP, but SNI does not support IP addreses
    * If the cluster is private, the AWS private link controller will set up two hypershift.local DNS entries for the router, one for the KAS, one for the apps domain

/label tide/merge-method-squash

enxebre commented on api/v1alpha1/endpointservice_types.go:
```diff
@@ -51,6 +51,10 @@ type AWSEndpointServiceStatus struct {
 	// +optional
 	DNSName string `json:"dnsName,omitempty"`
 
+	// DNSName are the names for the records created in the hypershift private zone
```
how is this different from the above field?

alvaroaleman replied:
It is different in that it can hold more than one value

enxebre replied:
so the semantic is the same, I don't we need two different fields to represent 1 vs multiple. Can we drop DNSName and keep only []DNSName where you can have 1..N?

alvaroaleman replied:
Yeah, the semantic is the same. We can deprecate the field, but we have to keep it around for or as long as we support an OCP version where the HCP controller does not have this change

enxebre commented:
>Making the CLI decide based on the presence of a value for --external-dns-domain if it configures Route or the existing LoadBalancer as publishingStrategy for the KAS

Why is external dns the discriminator, why don't always go router?

Can we please capture this in https://github.com/openshift/enhancements/tree/master/enhancements/hypershift/networking

alvaroaleman commented:
> Why is external dns the discriminator, why don't always go router?

Because it is not going to work without external DNS, as we need it to be able to do name based virtual hosting

csrwng commented on hypershift-operator/controllers/platform/aws/controller.go:
```diff
@@ -302,15 +302,15 @@ func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointServiceStatus(ctx con
 	hasPrivateIngressControllerEPService := false
 	for _, eps := range endpointServices.Items {
-		// If a private-router AWSEndpointService exists, it means that
+		// If a router AWSEndpointService exists, it means that
```
it means that ?

alvaroaleman replied:
No idea, I had just deleted the reference to private, whoever originally created the comment forgot to finish the sentence :upside_down_face:

csrwng reviewed (commented):
did an initial pass, testing to better understand new behavior

csrwng commented:
@alvaroaleman so in the public use case I would still expect the ignition, konnectivity, oauth and ovnsdb routes to go through the one router as long as a hostname is specified for them. That doesn't look to be the case with the current code.

alvaroaleman commented:
> @alvaroaleman so in the public use case I would still expect the ignition, konnectivity, oauth and ovnsdb routes to go through the one router as long as a hostname is specified for them. That doesn't look to be the case with the current code.

Updated that and added an e2e verification.The only exception for this is currently the ovn route, which is managed by the CNO. I will fix it there once this has merged.

alvaroaleman commented:
/retest-required

alvaroaleman commented:
/retest-required

alvaroaleman commented:
@csrwng @sjenning so in order to save on LB ports to in turn save on security group rules I added defaulting to listen on port 443 when the Route strategy is used. This required to:
* Change things to hardcode the kas to listen on 6443 and do the 443->6443 translation in the Service object, because the KAS is can not bind to 443 as it doesn't run as root
* Make the internal LB in private clusters respect the apiserver port, everything that uses it is hardcoded to use port 6443

Those two changes could be extracted from this PR, should I do that?

alvaroaleman commented:
/retest-required

csrwng commented:
> Those two changes could be extracted from this PR, should I do that?

Only if you think it will help to make progress quicker. I wouldn't spend too much time extracting it from here.

alvaroaleman commented:
Tests are finally green :tada:

enxebre commented on control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:
```diff
@@ -732,17 +740,13 @@ func (r *HostedControlPlaneReconciler) update(ctx context.Context, hostedControl
-	// Reconcile private router
+	// Reconcile router
 	if util.IsPrivateHCP(hostedControlPlane) {
 		r.Log.Info("Removing private IngressController")
```
nit: can we moved this up to line 560 block?

enxebre commented on control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:
```diff
@@ -818,10 +822,36 @@ func (r *HostedControlPlaneReconciler) reconcileAPIServerService(ctx context.Con
 	}
 
-	if util.IsPrivateHCP(hcp) {
+	if serviceStrategy.Type == hyperv1.Route {
```
nit: I'd find easier to read and understand all the possible casuistic with a swtich case for `serviceStrategy.Type` here, then in default with `util.IsPrivateHCP(hcp) {...`

alvaroaleman replied:
The issue is that we have three logical branches:
* Uses hyperv1. Router: Always create internal route.
    * subbranch is public: Create public route
* Uses hyperv1.LB and is private: Create private LB

The `hyperv1.LB` for the last condition is implicit, because private AWS clusters can only have route or LB apiserver expose strategy. I've added it explicitly to the condition to make that more obvious, the check doesn't really cost anything.

enxebre commented on control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:
```diff
@@ -556,6 +556,14 @@ func (r *HostedControlPlaneReconciler) update(ctx context.Context, hostedControl
+	kasServiceStrategy := servicePublishingStrategyByType(hostedControlPlane, hyperv1.APIServer)
+	if util.IsPrivateHCP(hostedControlPlane) || kasServiceStrategy.Type == hyperv1.Route {
+		r.Log.Info("Reconciling router")
+		if err = r.reconcileRouter(ctx, hostedControlPlane, releaseImage, createOrUpdate); err != nil {
```
why does this only happen when is private? I though part of the goal was "The per-hostedcluster router is always created, not only in the private case"

alvaroaleman replied:
I'll respond below

enxebre commented:
>Why is external dns the discriminator, why don't always go router?
Because it is not going to work without external DNS, as we need it to be able to do name based virtual hosting

Would it work if management cluster DNS is used when no-external DNS is set?

To keep solidifying and simplifying this, I'd expect:
- We agree on more opinionated valid setups and let the HC backend validate input and fail early in condition otherwise. E.g all components can be set to strategy router but not only some...
- Always create the router (regardless private or public setup).
- Let the CLI always set router strategy (regardless external DNS).

PR lgtm overall, dropped some nits and questions but no objections to proceed and follow up later.

alvaroaleman commented:
@csrwng could you have a look, please?

csrwng commented:
/hold cancel
/lgtm
